### PR TITLE
clientv3: document context to "Watch" API 

### PR DESCRIPTION
**Pull request by gyuho**
Address https://github.com/coreos/etcd/issues/8993.

### CODE REVIEW

File: clientv3/watch.go [hunk 1]
```diff
@@ -46,6 +46,20 @@ type Watcher interface {
 	// through the returned channel. If revisions waiting to be sent over the
 	// watch are compacted, then the watch will be canceled by the server, the
 	// client will post a compacted error watch response, and the channel will close.
+	// If the context "ctx" is canceled or timed out, returned "WatchChan" is closed,
+	// and "WatchResponse" from this closed channel has zero events and nil "Err()".
+	// If the context is "context.Background/TODO", returned "WatchChan" will not be closed
+	// and wait until events happen, except when server returns an error and:
```
Comment (xiang90) [inline clientv3/watch.go]: except when server returns non-recoverable error.

When require leader is passed in, watch will return `no leader` error when the connected server has no leader.

you probably need to polish my words, but you get the idea.

File: clientv3/watch.go [hunk 2]
```diff
@@ -46,6 +46,20 @@ type Watcher interface {
...
+	//  1. error is "etcdserver: no leader", then "WatchChan" is closed with non-nil "Err()".
+	//  2. error is "grpc/*status.statusError" and its error code is neither codes.Unavailable
+	//     nor codes.Internal, then "WatchChan" is closed with non-nil "Err()".
+	// If any of above cases is not "true" and context has not been canceled or timed out,
```
Comment (xiang90) [inline clientv3/watch.go]: watch will retry recoverable errors forever until reconnected.

**Comment by gyuho** [comment]
@xiang90 PTAL.

File: clientv3/watch.go [hunk 3]
```diff
@@ -46,6 +46,24 @@ type Watcher interface {
...
 	// client will post a compacted error watch response, and the channel will close.
+	// If the context "ctx" is canceled or timed out, returned "WatchChan" is closed,
+	// and "WatchResponse" from this closed channel has zero events and nil "Err()".
+	// If the context is "context.Background/TODO", returned "WatchChan" will
+	// not be closed and wait until events happen, except when server returns
+	// a non-recoverable error and:
+	//  1. When context passed with "WithRequireLeader" and the connected server
+	//     has no leader, error "etcdserver: no leader" is returned, and then
+	//     "WatchChan" is closed with non-nil "Err()".
+	//  2. error is "grpc/*status.statusError" and its error code is neither
```
Comment (xiang90) [inline clientv3/watch.go]: isnt this covered by non-recoverable error? i do not want to list all errors here.

Comment (gyuho) [reply clientv3/watch.go]: @xiang90 Just removed that part. PTAL.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**